modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -133,90 +138,377 @@ public void initCluster(Collection<String> metaStorageNodeNames, Collection<Stri
+                            }));
+        }
+
+        resultHook.whenComplete((v, e) -> {
+            NetworkMessage response = e == null
+                    ? msgFactory.initCompleteMessage().build()
+                    : msgFactory.initErrorMessage().cause(e.getMessage()).build();
+
+            clusterService.messagingService().respond(addr, response, correlationId);
+        });
+    }
+
+    /**
+     * Starts the CMG Raft service and writes the given {@code state} to the storage.
+     */
+    private CompletableFuture<CmgRaftService> initCmgRaftService(ClusterState state) {
+        return localStateStorage.putCmgNodeNames(state.cmgNodes())
+                .thenCompose(v -> startCmgRaftService(state.cmgNodes()))
+                .thenCompose(service -> service.isCurrentNodeLeader()
+                        .thenCompose(isLeader -> isLeader ? initCmgState(service, state) : completedFuture(null))
+                        .thenApply(v -> service));
+    }
+
+    /**
+     * Writes the given state to the CMG's STM and executes some necessary on-leader logic.
+     */
+    private CompletableFuture<Void> initCmgState(CmgRaftService service, ClusterState state) {
+        return service.writeClusterState(state).thenCompose(v -> onLeaderElected(service, state));
+    }
+
+    /**
+     * Executes the following actions when a CMG leader is elected.
+     * <ol>
+     *     <li>Updates the logical topology in case some nodes have gone offline during leader election.</li>
+     *     <li>Broadcasts the current CMG state to all nodes in the physical topology.</li>
+     * </ol>
+     */
+    private CompletableFuture<Void> onLeaderElected(CmgRaftService service, ClusterState state) {
+        return updateLogicalTopology(service)
+                .thenRun(() -> {
+                    clusterService.topologyService().addEventHandler(cmgLeaderTopologyEventHandler(service));
+
+                    sendClusterState(state, clusterService.topologyService().allMembers());
+                });
+    }
+
+    /**
+     * This method must be executed upon CMG leader election in order to regain logical topology consistency in case some nodes left the
+     * physical topology during the election. New node will be added automatically after the new leader broadcasts the current cluster
+     * state.
+     */
+    private CompletableFuture<Void> updateLogicalTopology(CmgRaftService service) {
+        return service.logicalTopology()
+                .thenAccept(logicalTopology -> {
+                    Collection<String> physicalTopologyIds = clusterService.topologyService().allMembers()
+                            .stream()
+                            .map(ClusterNode::id)
+                            .collect(toSet());
+
+                    for (ClusterNode node : logicalTopology) {
+                        if (!physicalTopologyIds.contains(node.id())) {
+                            removeNodeFromLogicalTopology(service, node);
```

Comment:
> I suggest renaming the method

done

> Also, why should it be done in an async way, using a scheduled task?

This is needed in case a node is lost temporarily (due to a network glitch, for example). We want to avoid node re-validation and data re-balancing in this case. In other words, we give the node a change to get back into the logical topology.